### ISSUE TITLE
[ENHANCEMENT]Improve support for multiple output media types

### ISSUE THREAD

**Issue by axtsnee** [softwaremill/tapir#1502]
Tapir version: 0.18.3

Scala version: 2.13.6

**Current state**

Right now tapir has excellent support for API versioning by URL. API versioning by content type is possible but less easy. For example, say I want to have one endpoint take a single input type but return one of two different output types based on Accept header content negotiation. Right now I have to describe that endpoint like this:
```scala
  val v1mediaType = MediaType("application", "vnd.com.example.output.v1+json")
  val v2mediaType = MediaType("application", "vnd.com.example.output.v2+json")
  def customBody[T](mt: MediaType)(implicit ev: Codec[String, T <: VersionedOutputTrait, CodecFormat.Json]): EndpointIO.Body[String, VersionedOutputTrait] =
    anyFromUtf8StringBody(ev.format(new CodecFormat {
      override def mediaType: MediaType = mt
    }))
  val versionedServerEndpoint: ServerEndpoint[(MediaType, InputCaseClass), Throwable, VersionedOutputTrait, F] =
    baseEndpoint.post
      .in(acceptHeader(
        v1mediaType,
        v2mediaType,
        MediaType.ApplicationJson
      ))
      .in("versioned")
      .in(jsonBody[InputCaseClass])
      .out(
        oneOf(
          oneOfMapping(StatusCode.Ok, customBody[VersionedOutputCaseClass1](v1mediaType)),
          oneOfMapping(StatusCode.Ok, customBody[VersionedOutputCaseClass2](v2mediaType))
        )
      ).serverLogic {
        case (v1mediaType, input) => serviceV1.logic(input) // returns F[VersionedOutputCaseClass1]
        case (_, input) => serviceV2.logic(input) // returns F[VersionedOutputCaseClass2]
      }
``` 
where `VersionedOutputCaseClass1` and `VersionedOutputCaseClass2` have to extend marker trait `VersionedOutputTrait`. Furthermore, that trait's companion object must have an implicit JSON encoder for `VersionedOutputTrait` that delegates to the encoders for `VersionedOutputCaseClass1` and `VersionedOutputCaseClass2`. Finally, `acceptHeader` is a function with parameters `MediaType*` and output `EndpointInput[MediaType]` that has sophisticated logic to parse the Accept header and return the MediaType parameter that matches best.

I do not yet have a workaround for the problem with the generated OpenAPI documentation, which contains
```YAML
      responses:
        '200':
          description: ''
          content:
            application/vnd.com.example.output.v1+json:
              schema:
                $ref: '#/components/schemas/VersionedOutputTrait'
            application/vnd.com.example.output.v2+json:
              schema:
                $ref: '#/components/schemas/VersionedOutputTrait'
...
components:
  schemas:
    VersionedOutputTrait:
      oneOf:
      - $ref: '#/components/schemas/VersionedOutputCaseClass1'
      - $ref: '#/components/schemas/VersionedOutputCaseClass2'
...
```
This YAML does not give SwaggerUI enough information to match content type `application/vnd.com.example.output.v1+json` to schema `#/components/schemas/VersionedOutputCaseClass1`.

**Desired state**
```scala
  val versionedServerEndpoint: ServerEndpoint[InputCaseClass, Throwable, _, F] =
    baseEndpoint.post
      .in("versioned")
      .in(jsonBody[InputCaseClass])
      .out(
        oneOf(
          oneOfMapping(StatusCode.Ok, jsonBody[VersionedOutputCaseClass1]),
          oneOfMapping(StatusCode.Ok, jsonBody[VersionedOutputCaseClass2])
        )
      ).serverLogic {
        case (???, input) => serviceV1.logic(input) // returns F[VersionedOutputCaseClass1]
        case (_, input) => serviceV2.logic(input) // returns F[VersionedOutputCaseClass2]
      }
``` 

```YAML
      responses:
        '200':
          description: ''
          content:
            application/vnd.com.example.output.v1+json:
              schema:
                $ref: '#/components/schemas/VersionedOutputCaseClass1'
            application/vnd.com.example.output.v2+json:
              schema:
                $ref: '#/components/schemas/VersionedOutputCaseClass2'
...
```

**Comment by adamw** [comment]
Btw. as part of another feature, I'll want to simplify 

```scala
        oneOf(
          oneOfMapping(StatusCode.Ok, customBody[VersionedOutputCaseClass1](v1mediaType)),
          oneOfMapping(StatusCode.Ok, customBody[VersionedOutputCaseClass2](v2mediaType))
        )
``` 

into

```scala
        oneOf(
          oneOfMapping(customBody[VersionedOutputCaseClass1](v1mediaType)),
          oneOfMapping(customBody[VersionedOutputCaseClass2](v2mediaType))
        )
```

and allow multi-level `oneOf`s. This might help with your example as well.

**Comment by axtsnee** [comment]
Thank you for responding.

I am not sure about the statement "single tapir endpoint <-> single openapi endpoint". As I understand it, if endpoint with path `/x` supports GET, POST, and DELETE, there need to be three tapir endpoints for one openapi endpoint. So `path` is not unique - `(path, method)` is unique. As long as there is only one server logic per tapir endpoint, it has to be that way, does it not?

So it seems there are two proposals to choose from. I proposed making `(path, method, media type)` unique, and keeping the correspondence single tapir endpoint <-> single server logic. The alternative is to keep `(path, method)` unique but somehow have multiple server logics. Something like this?
```scala
    .serverLogic(
      oneOf(
        oneOfMapping(v1mediaType, input => serviceV1.logic(input)),
        oneOfMapping(v2mediaType, input => serviceV2.logic(input))
      )
    )
```

**Comment by adamw** [comment]
Yes, and in openapi for `GET /x` and `/POST /x` you'll also have separate entries in the yaml file. While not a strict requirement - you can have tapir endpoints with the same method & path which discriminate on some other property, if you change the `DecodeFailureHandler` - by default `(path, method)` uniquely identify an endpoint.

Isn't it that in most cases with multiple media types the server logic will be the same, except for output formatting?

Also I think there would be another problem in making `(path, method, media type)` unique. Path & method are properties of the input, while the media type is a property of the output - and while it can be influenced by `Accept` header, it can be dynamically generated as well. There are also other output properties which can vary, such as status codes. So I think the current `oneOf` approach scales better

**Comment by adamw** [comment]
> then you have to allow multiple input types (which tapir does very well) AND multiple output types

I think it's the other way round, in a single endpoint, it's easy to have multiple output types, but not input types? The only way to have multiple input types is to create a json decoder for a trait and discriminate there.

Anyway, I get your point. And maybe it wouldn't be that hard to combine endpoints with same path+method in the OpenAPI interpreter, given that their content types don't overlap. However, what about other inputs? What if there's a query parameter is one endpoint, and absent in another? This would require quite a lot of discipline to properly define the endpoints. That's why I'd be leaning towards finding a solution for the api-versioning problem using content types, but using a single endpoint. What do you think?

**Comment by adamw** [comment]
While I agree that not ideal, sticking with the endpoint <-> path & method correspondence seems pragmatic: having to combine multiple endpoints into a  single OpenAPI operation would be risky and probably buggy. For this particular case, the  relatively new [`oneOfBody`](https://tapir.softwaremill.com/en/latest/endpoint/oneof.html#oneofbody-inputs-outputs) might be helpful, though the problem of matching input & output versions remains and will still need manual coding.
